include/msgpack/sbuffer.h

```diff
@@ -81,7 +81,7 @@ static inline int msgpack_sbuffer_write(void* data, const char* buf, size_t len)
         sbuf->alloc = nsize;
     }
 
-    memcpy(sbuf->data + sbuf->size, buf, len);
+    if(buf) { memcpy(sbuf->data + sbuf->size, buf, len); }
     sbuf->size += len;
```

Comment:
Should this be inside the `if`?

Reply:
I think that the code is no problem.
Source data are `buf` and `len`. And the destination is `sbuf->data`.

If `buf` is nullptr then doesn't copy. This archives nullptr dereference.
Then simply add size.

Reply:
I prefer to be paranoid. If `buf` is null but `len` is non-zero (a bug in the calling code), why should `sbuf->size` change?

Reply:
I think that it is msgpack-c API spec definition problem.
The error hannpens at
https://github.com/msgpack/msgpack-c/blob/82ddfc11709211c43fe2e8820934f3d95b2b6056/test/msgpack_c.cpp#L875

str is NULL.

```c
    msgpack_pack_str_body(&pk, str, str_size);
```

msgpack_pack_str_body() accepts or doesn't accept NULL 2nd parameter is not documented.

If msgpack_pack_str_body() accept it, how str_size is treated. It is also not documented.

Possible choice:
1. Ignore str_size.
2. Advance the pointer by str_size without copy.

(3. If msgpack_pack_str_body() doesn't accept NULL str, then UB.)

I think that 3 is not good. 
The current PR is 2. 
If msgpack_pack_str_body() respects str, then 1 is better.
If msgpack_pack_str_body() respects str_size, then 2 is better.

I think that choosing 1 or 2 is just a preference.

I think that `assert(str || str_size == 0);` should be added on the top of the function.
The assert means if str is NULL then str_size should be 0. If str is not NULL, any str_size is OK including 0.

Reply:
@ygj6, could you add assert ?

And, it is out of spec but could you update the code as folows ?

```c
    if(buf) { 
        memcpy(sbuf->data + sbuf->size, buf, len); 
        sbuf->size += len;
    }
}
```

I choose option 1 above.

BTW, the real problem is poor documentation. I know about that.
Unfortunately I don't have much time to write Doxygen style document (comment) like 
https://github.com/msgpack/msgpack-c/blob/2f5b13799743f800655680b8eab327c739aad2b5/include/msgpack/v1/unpack.hpp#L933-L947
 to whole code.

So I wrote minimal document on wiki.
https://github.com/msgpack/msgpack-c/wiki/v2_0_c_overview

We will add document to wiki or code in the future.

Reply:
Okay, I will add `assert(str || str_size == 0)` on line 875 in `msgpack-c/test/msgpack_c.cpp`.

Reply:
@ygj6 ,sorry to confuse you.
 I mean adding assert to 
https://github.com/msgpack/msgpack-c/blob/ba15089d8601bb08155fe8fbb90b49a10728bfaf/include/msgpack/sbuffer.h#L61
.

The combination of NULL str and non zero len is apparently misuse. So I think that assert should be added in the library.
In the release build, assert is simple removed. For this case, 

```c
    if(buf) { 
        memcpy(sbuf->data + sbuf->size, buf, len); 
        sbuf->size += len;
    }
```

avoid copy and increment size.

Reply:
For some older compilers, the statement must follow the variable declaration. Therefore, I will add `assert(buf || len == 0)` after the line `msgpack_sbuffer* sbuf = (msgpack_sbuffer*)data;`. What do you think?

Reply:
> For some older compilers, the statement must follow the variable declaration. Therefore, I will add `assert(buf || len == 0)` after the line `msgpack_sbuffer* sbuf = (msgpack_sbuffer*)data;`. What do you think?

I agree. I forgot about oder C (ANSI-C).